cmd/main.go

```diff
@@ -37,9 +48,23 @@ func main() {
 	backendServer := backend.NewServerWithJSONRPC(jsonRPC)
 	middleendServer := middleend.NewServerWithJSONRPC(jsonRPC)
 
-	pb.RegisterFrontendNvmeServiceServer(s, frontendServer)
-	pb.RegisterFrontendVirtioBlkServiceServer(s, frontendServer)
-	pb.RegisterFrontendVirtioScsiServiceServer(s, frontendServer)
+	// TODO: Move interface to frontend and work with it rather than concrete types.
+	type feOpiServer interface {
+		pb.FrontendNvmeServiceServer
+		pb.FrontendVirtioBlkServiceServer
+		pb.FrontendVirtioScsiServiceServer
+	}
+	var frontendOpiServer feOpiServer
+	if useKvm {
+		log.Println("Creating KVM server.")
+		frontendOpiServer = feOpiServer(kvm.NewKvmServer(frontendServer, qmpAddress, ctrlrDir))
+	} else {
+		frontendOpiServer = feOpiServer(frontendServer)
+	}
+
+	pb.RegisterFrontendNvmeServiceServer(s, frontendOpiServer)
+	pb.RegisterFrontendVirtioBlkServiceServer(s, frontendOpiServer)
+	pb.RegisterFrontendVirtioScsiServiceServer(s, frontendOpiServer)
```

Comment:
why do we need new interface ?
why not something more simple like this:
```suggestion

kvmServer := kvm.NewKvmServer(frontendServer, qmpAddress, ctrlrDir)

if useKvm {
	pb.RegisterFrontendNvmeServiceServer(s, kvmServer)
	pb.RegisterFrontendVirtioBlkServiceServer(s, kvmServer)
	pb.RegisterFrontendVirtioScsiServiceServer(s, kvmServer)
} else {
	pb.RegisterFrontendNvmeServiceServer(s, frontendServer)
	pb.RegisterFrontendVirtioBlkServiceServer(s, frontendServer)
	pb.RegisterFrontendVirtioScsiServiceServer(s, frontendServer)
}
```

Reply:
Because of the code duplication. It is small, but sometimes it points out that we are missing something. If we pass different types to `Register*` methods, it means those types already have required interfaces, we just need to extract one. In addition, I expect we need that interface, if we want to move our test infrastructure from different packages into utils.go (otherwise we have a circular dependency) Also interfaces do not allow accessing data from the original type, so a better encapsulation is expected as well

Reply:
I think readability is more important in this specific case
we can always complicate later
for this PR I prefer a simpler solution with simple if-else and well defined types without additional future looking abstraction for now
maybe next PRs will show clearly the need, but for now I don't see it